packages/adapters/watcher/src/adapter.ts

```diff
@@ -151,4 +170,51 @@ export class WatcherAdapter {
       throw errors;
     }
   }
+
+  /**
+   * @notice Resets the latest transactions to ensure we only capture
+   * ones relevant to intervals of invariant calls
+   */
+  private clearTransactions(): void {
+    this.context.domains.forEach((domain) => {
+      this.latestTransactions[domain] = [];
+    });
+  }
+
+  /**
+   * @notice Stores transactions at the first instance of invariant failures
+   */
+  private recordTransactions(): void {
+    if (this.transactionSnapshot) {
+      // Do not override, likely already set from previous round of alerting
+      return;
+    }
+    // Store transactions sent to connext in latest interval
+    this.transactionSnapshot = { ...this.latestTransactions };
+  }
+
+  /**
+   * @notice Kicks off a listener to log any transaction hash that has been sent to connext
+   * between the `checkInvariant` calls. This allows us to log any transactions that may
+   * have caused the invariant failure immediately.
+   */
+  private beginRecordingTransactions(): void {
+    // clear out all the existing transactions. also ensures an array will always be present
+    this.clearTransactions();
+    // start listeners for each transaction
+    this.context.domains.forEach((_domain) => {
+      const domain = +_domain;
+      const chain = domainToChainId(domain);
+      const provider = this.context.txservice.getProvider(domain);
+      // FIXME: the only way to access events is by using the fallback provider, which
+      // will soon be deprecated
```

Comment:
alternatively you can have an ethers Contract instance that has attached listeners
https://docs.ethers.org/v5/api/contract/contract/#Contract-listenerCount
Will technically be inefficient as we'll have a redundant provider nested in the instance, but at least it won't be a temporary solution / obsoleted in the future